File: frontend/src/services/certificate-service.ts
import { httpClient } from '../lib';

export type CertificateStatus = 'ISSUED' | 'REVOKED';

export type CertificateVerifyResponse = {
  credentialId: string;
  status: CertificateStatus;
  studentName: string;
  courseTitle: string;
  courseThumbnailUrl?: string | null;
  instructorName: string;
  issuedAt: string;
  revokedAt?: string | null;
  revokeReason?: string | null;
};

export type CertificateResponse = {
  id: string;
  studentId: string;
  courseId: string;
  credentialId: string;
  status: CertificateStatus;
  studentName: string;
  courseTitle: string;
  courseThumbnailUrl?: string | null;
  instructorName: string;
  issuedAt: string;
  revokedAt?: string | null;
};

export type CertificateDownloadResponse = {
  url: string;
};

export type AdminCertificatePageResponse = {
  items: CertificateResponse[];
  totalRecords: number;
  totalPages: number;
  page: number;
  size: number;
  totalIssued: number;
  totalRevoked: number;
};

export async function verifyCertificate(code: string) {
  const response = await httpClient.get<CertificateVerifyResponse>(
    `/certificate/verify/${encodeURIComponent(code)}`,
    { skipAuthRefresh: true },
  );
  return response.data;
}

export async function getMyCertificates() {
  const response = await httpClient.get<CertificateResponse[]>('/student/certificates');
  return response.data;
}

export async function getMyCertificate(id: string) {
  const response = await httpClient.get<CertificateResponse>(
    `/student/certificates/${encodeURIComponent(id)}`,
  );
  return response.data;
}

export async function getCertificateDownloadUrl(id: string) {
  const response = await httpClient.get<CertificateDownloadResponse>(
    `/student/certificates/${encodeURIComponent(id)}/download`,
  );
  return response.data;
}

export async function getAdminCertificates(params?: { page?: number; size?: number; search?: string; status?: string }) {
  const response = await httpClient.get<AdminCertificatePageResponse>('/admin/certificates', { params });
  return response.data;
}

export async function revokeCertificate(id: string, reason: string) {
  const response = await httpClient.post<CertificateResponse>(
    `/admin/certificates/${encodeURIComponent(id)}/revoke`,
    { reason },
  );
  return response.data;
}

File: backend/src/main/java/project/lms_rikkei_edu/modules/certificate/service/CertificateService.java
package project.lms_rikkei_edu.modules.certificate.service;

import project.lms_rikkei_edu.modules.certificate.dto.request.RevokeCertificateRequest;
import project.lms_rikkei_edu.modules.certificate.dto.response.AdminCertificatePageResponse;
import project.lms_rikkei_edu.modules.certificate.dto.response.CertificateDownloadResponse;
import project.lms_rikkei_edu.modules.certificate.dto.response.CertificateResponse;
import project.lms_rikkei_edu.modules.certificate.dto.response.CertificateVerifyResponse;

import java.util.List;
import java.util.UUID;

public interface CertificateService {

    void issueCertificateIfEligible(UUID studentId, UUID courseId);

    List<CertificateResponse> getAllCertificatesForAdmin();

    AdminCertificatePageResponse getAdminCertificates(int page, int size, String search, String status);

    List<CertificateResponse> getMyCertificates(UUID studentId);

    CertificateResponse getMyCertificate(UUID studentId, UUID certificateId);

    CertificateDownloadResponse getStudentDownloadUrl(UUID studentId, UUID certificateId);

    CertificateVerifyResponse verify(String credentialId);

    CertificateResponse revoke(UUID adminId, UUID certificateId, RevokeCertificateRequest request);
}
